Lib/socket.py

```diff
@@ -728,6 +728,78 @@ def create_connection(address, timeout=_GLOBAL_DEFAULT_TIMEOUT,
     else:
         raise error("getaddrinfo returns an empty list")
 
+def bind_socket(address, family=AF_UNSPEC, type=SOCK_STREAM, *, backlog=128,
+                reuse_port=False, flags=getattr(_socket, "AI_PASSIVE", 0)):
+    """Convenience function which creates a socket bound to *address*
+    (a 2-tuple (host, port)) and return the socket object.
+
+    If *host* is an empty string or None all network interfaces are assumed.
+
+    If *family* is AF_UNSPEC or None the address family will be
+    determined from the *host* specified in *address*.
+
+    *type* should be either SOCK_STREAM or SOCK_DGRAM.
+
+    *backlog* is the queue size passed to socket.listen() and is ignored
+    for SOCK_DGRAM socket types.
+
+    *flags* is a bitmask for getaddrinfo().
+
+    >>> with bind_socket((None, 8000)) as server:
+    ...     while True:
+    ...         conn, addr = server.accept()
+    ...         # handle new connection
+    """
+    host, port = address
+    if host == "":
+        # https://mail.python.org/pipermail/python-ideas/2013-March/019937.html
+        host = None  # all interfaces
+    if not family:
+        family = AF_UNSPEC
+    reuse_addr = hasattr(_socket, 'SO_REUSEADDR') and \
```

Comment:
`asyncio` has this logic only if `reuse_addr` parameter is `None` (default).
It allows to explicitly enforce and *disable* reuse addr feature.
I've found the logic useful for some cases.

Reply:
@asvetlov I'm *extremely* curious what situation you've found where you don't want `SO_REUSEADDR` :-).

As far as I understand it, `SO_REUSEADDR` is simply always wanted on POSIX; it's a classic "unbreak-me" flag. I don't think there's any harm in allowing people to disable it, I'm just really curious to hear why anyone would want to do this :-).

On Windows, I have a stronger opinion: we should not provide any way to control `SO_REUSEADDR` on Windows. On Windows, the behavior that most people expect `SO_REUSEADDR` to control (i.e., binding to ports that are in the TIME_WAIT state) is always enabled, and cannot be configured. But if you use `SO_REUSEADDR` anyway, then this creates a security bug (and also a regular bug). It would be irresponsible to provide a flag where `False` does nothing and `True` breaks security. Simply making the default `False` on Windows is not enough; using `SO_REUSEADDR` should be impossible.

On Windows you should also enable `SO_EXCLUSIVEADDRUSE`, which is required to fully fix the security bug I mentioned. It has no other effect except to fix this bug.

Reply:
As for `SO_REUSEADDR` this excellent SO reply explains why disabling `SO_REUSEADDR` can be desirable (although not frequently): https://stackoverflow.com/a/14388707. In short, when not set it results in `EADDRINUSE` error if:
1) a wildcard address ("all" interfaces) is specified as "host" and some address/port is already bound to another socket
2) there is some pending data to send
...so it appears the argument/toggle is desirable.

As for `SO_REUSEADDR` it seems the semantic is different on Windows. When set to `1` it forces the socket to forcibly bind to a port in use by another socket, whereas on UNIX this results in `EADDRINUSE` error. It also sets the process in a wedged state requiring where it'll no longer respond to any signals or graceful kills, and must be forcibly killed via OpenProcess()/TerminateProcess(). 
Sources: [test.support](https://github.com/python/cpython/blob/7ab3d1573c123fdd582a239d01f8475651df38f2/Lib/test/support/__init__.py#L652), [BPO-2550], (https://bugs.python.org/issue2550). So I agree `reuse_addr=True` should be ignored on Windows (and documented).

As for `SO_EXCLUSIVEADDRUSE` (Windows), I'm dubious. MS doc states:

>  a socket with the SO_EXCLUSIVEADDRUSE set cannot necessarily be reused immediately after socket closure. For example, if a listening socket with SO_EXCLUSIVEADDRUSE set accepts a connection and is then subsequently closed, another socket (also with SO_EXCLUSIVEADDRUSE) cannot bind to the same port as the first socket until the original connection becomes inactive.

...so it seems this is not desirable to set by default or change the meaning of `reuse_addr` opt on Windows to mean `SO_EXCLUSIVEADDRUSE` instead of `SO_REUSEADDR`. Also, I took a peek at Tornado, Twisted, asyncio, curio and none of them use `SO_EXCLUSIVEADDRUSE`.

Reply:
Yes, I'm pretty sure those MS docs are why none of the other libraries use it. However, the docs are wrong :-). [Try it](https://github.com/python-trio/trio/blob/master/notes-to-self/time-wait-windows-exclusiveaddruse.py) – Windows is totally happy to let you `bind` a new listening socket to a port that still has live connections from a previous listening socket.

Reply:
I wouldn't be so sure about [this script](https://github.com/python-trio/trio/blob/master/notes-to-self/time-wait-windows-exclusiveaddruse.py). It appears the "rebind with TIME_WAIT socket" test succeeds regardless of `SO_EXCLUSIVEADDRUSE` (I tried).

The whole point of `reuse_addr=True` on UNIX is to avoid getting `EADDRINUSE` when the previous socket is still in `TIME_WAIT` state. AFAIU WinSock behaves like that already without setting any option. For this reason I think it's correct to not set `SO_REUSEADDR` by default on Windows, nor to map `reuse_addr=True` to `SO_EXCLUSIVEADDRUSE` because fundamentally `SO_REUSEADDR` and `SO_EXCLUSIVEADDRUSE` have different meanings.

Also I think I changed my mind about disallowing to set `SO_REUSEADDR` on Windows because despite the security implications it appears there is a use-case for it. From MS doc:

> (about SO_REUSERADDR) The exception to this non-deterministic behavior is multicast sockets. If two sockets are bound to the same interface and port and are members of the same multicast group, data will be delivered to both sockets, rather than an arbitrarily chosen one.
> [...]
> The SO_REUSEADDR option has very few uses in normal applications aside from multicast sockets where data is delivered to all of the sockets bound on the same port.

As such I prefer the user to be able to set/unset it if he/she wants to.

Reply:
> I wouldn't be so sure about this script. It appears the "rebind with TIME_WAIT socket" test succeeds regardless of SO_EXCLUSIVEADDRUSE (I tried).

Dunno why you're "not so sure" about the script. It appears to have done exactly what I claimed.

> For this reason I think it's correct to not set SO_REUSEADDR by default on Windows, nor to map reuse_addr=True to SO_EXCLUSIVEADDRUSE because fundamentally SO_REUSEADDR and SO_EXCLUSIVEADDRUSE have different meanings.

I agree. That's why I never suggested doing this. This is a very odd conversation.

Again, my suggestion is: on Windows, use SO_EXCLUSIVEADDRUSE and don't use SO_REUSEADDR. That's all.

> Also I think I changed my mind about disallowing to set SO_REUSEADDR on Windows because despite the security implications it appears there is a use-case for it.

I disagree. I don't think this high-level convenience function should provide a giant security footgun in order to support one very rare use case. If that use case is important enough to get a high-level convenience function, then it's important enough to make that function actually convenient.

I'll leave it to @1st1 and @asvetlov decide whether they want to enforce this before merging.

Reply:
> @asvetlov I've found the logic useful for some cases.

I'm also curious what were those cases.

> I'll leave it to @1st1 and @asvetlov decide whether they want to enforce this before merging.

After following this discussion I tend to side with Nathaniel's line of reasoning. The use-case for not setting `SO_REUSEADDR` is indeed extremely narrow. I strongly suggest to simply drop the `reuse_addr` parameter (and set `SO_REUSEADDR` always on *nix).

Reply:
> Again, my suggestion is: on Windows, use SO_EXCLUSIVEADDRUSE and don't use SO_REUSEADDR. That's all.

My current objections against using `SO_EXCLUSIVEADDRUSE` by default are:
1) [MS doc](https://docs.microsoft.com/en-us/windows/desktop/WinSock/using-so-reuseaddr-and-so-exclusiveaddruse) states <<Conversely, a socket with the SO_EXCLUSIVEADDRUSE set cannot necessarily be reused immediately after socket closure.>> which is precisely what we want to avoid here
2) setting no options at all already appears to match the same semantic of `SO_REUSEADDR` on UNIX (aka, return `EADDRINUSE` on second `bind()`). 
3) it's not being used by any major net framework I'm aware of and also by googling for `filetype:py SO_EXCLUSIVEADDRUSE` I get very few results. 
4) as for your script, I commented lines setting `SO_EXCLUSIVEADDRUSE` (23, 38, 47, 63) and on Windows 10 it prints exactly the same output. As such it appears it's probably not testing what was intended for? Is it "stealing an already used port" the security scenario you're afraid of? If so please paste a script able to reproduce it.

Reply:
I found a 6 years old issue where this was already discussed, basically coming up with the same considerations/conclusions: 
https://bugs.python.org/issue8184

Reply:
> > I'll leave it to @1st1 and @asvetlov decide whether they want to enforce this before merging.

> After following this discussion I tend to side with Nathaniel's line of reasoning. The use-case for not setting `SO_REUSEADDR` is indeed extremely narrow. I strongly suggest to simply drop the `reuse_addr` parameter (and set `SO_REUSEADDR` always on *nix).

Yes, I agree this is not worth it. Also the fact that `reuse_addr` would have a different semantic on UNIX vs. Windows for UDP sockets is a bad sign.

Reply:
>4. as for your script, I commented lines setting `SO_EXCLUSIVEADDRUSE` (23, 38, 47, 63) and on Windows 10 it prints exactly the same output. As such it appears it's probably not testing what it was intended for?

No, the point of the script is to demonstrate that settings `SO_EXCLUSIVEADDRUSE` doesn't break normal usage. It creates the situations where MSDN says using `SO_EXCLUSIVEADDRUSE` will cause problems, and demonstrates that in fact `SO_EXCLUSIVEADDRUSE` doesn't affect them at all. I haven't written a script to check that it actually fixes the bug/security hole it's supposed to fix. I guess that would be a good thing to do.

I believe the case where it's supposed to make a difference is:

Server A binds a socket. Then Server B attempts to bind the same address, and Server B uses `SO_REUSEADDR`. Supposedly, if Server A use `SO_EXCLUSIVEADDRUSE`, then Server B will get an error "socket already in use", but if Server A *doesn't* use `SO_EXCLUSIVEADDRUSE`, then Server B will hijack Server A's socket, and possibly cause other havoc. (The docs just say you get "undefined behavior".)

Reply:
> No, the point of the script is to demonstrate that settings `SO_EXCLUSIVEADDRUSE` doesn't break normal usage. It creates the situations where MSDN says using `SO_EXCLUSIVEADDRUSE` will cause problems, and demonstrates that in fact `SO_EXCLUSIVEADDRUSE` doesn't affect them at all.

If it doesn't affect them at all then why use it in the first place? It looks like the point of your script is to prove that MS doc is wrong. You are not even making a real case for it AFAICT.

> I believe the case where it's supposed to make a difference is: [this]

No. I've already explained why this is not the case multiple times: WinSock already do this by default. Why don't you verify your claims first?

```python
from socket import *
addr = ("localhost", 8888)
s1 = socket()
s1.bind(addr)
s2 = socket()
s2.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)  # will not cause port stealing
s2.bind(addr)  # fails with 10048 
```

Reply:
Wow. So now you're saying that the MSDN docs are wrong, and `SO_REUSEADDR` does not actually cause a security problem. That's super interesting if true! Thanks for pointing it out. However, what you are saying now is absolutely not what you said before. Quoting you:

> the semantic is different on Windows. When set to 1 it forces the socket to forcibly bind to a port in use by another socket, whereas on UNIX this results in EADDRINUSE error. It also sets the process in a state where it'll no longer respond to any signals or graceful kills, and must be forcibly killed (see test.support and BPO-2550).

> I think I changed my mind about disallowing to set `SO_REUSEADDR` on Windows because despite the security implications it appears there is a use-case for it. [→ so this is saying, you think that there are security implications]

This "already explained why this is not the case multiple times" thing you're claiming now is pure gaslighting.

> If it doesn't affect them at all then why use it in the first place? It looks like the point of your script is to prove that MS doc is wrong. You are not even making a real case for it AFAICT.

I have been straightforward and clear throughout about what I knew, what I suspected, and what my rationale was. I've responded to all of the points you've brought up, and repeated myself several times as you kept twisted what I said around into some unrelated strawman. And now you're being a patronizing jerk too.

I do care about the quality of the `socket` module, but I don't care enough to put up with your behavior. Have fun. Thanks for pointing out the `SO_REUSEADDR` thing; I'll probably investigate that more at some point. I'm unsubscribing from this thread now. Please don't tag me in again.

Reply:
@njsmith 
> Wow. So now you're saying that the MSDN docs are wrong, and `SO_REUSEADDR` does not actually cause a security problem.

That's not what I'm saying at all. We are all agreeing `SO_REUSEADDR` should not be used (nor allowed) on Windows because of the security implications. I even **quoted you** and agreed with you already that we shouldn't have a `reuse_addr` parameter anymore:
https://github.com/python/cpython/pull/11784#discussion_r256654410
So you either completely lost track of the conversation or you just want to provoke me. 

The point we're discussing about now is not `SO_REUSEADDR`. It's whether to use `SO_EXCLUSIVEADDRUSE` as *you*' (not me) have been claiming we should do all along. And you think we should because you erroneously believe that if we don't, a second socket setting `SO_REUSEADDR` may steal the port of the first one (your words):

 > Server A binds a socket. Then Server B attempts to bind the same address, and Server B uses `SO_REUSEADDR`. Supposedly, if Server A use `SO_EXCLUSIVEADDRUSE`, then Server B will get an error "socket already in use", but if Server A _doesn't_ use `SO_EXCLUSIVEADDRUSE`, then Server B will hijack Server A's socket, and possibly cause other havoc. (The docs just say you get "undefined behavior".)

But the code I showed you [above](https://github.com/python/cpython/pull/11784#discussion_r256673038) proves this is not the case and that you're wrong. This is what I've been trying to explain to you all along but never succeeded to. You've also never addressed my 4 arguments for **not** using `SO_EXCLUSIVEADDRUSE`, especially the number 1:
https://github.com/python/cpython/pull/11784#discussion_r256575317
So again, not even an apparent lack of understanding on your part, but a refusal to fully engage in a rational debate.  Which can be tolerable to some extent. When it stops being fine is the moment the discussion stops being technical and you start calling me names. Now this:

> This is a very odd conversation [...] Wow! That's super interesting if true! [...] And now you're being a patronizing jerk too. [...]  I'm unsubscribing from this thread now. [...] I don't care enough to put up with your behavior [...] Please don't tag me in again.

 ...is something I don't like and I'm not willing to accept. Just so you know.

> This is a very odd conversation [...] Wow! That's super interesting if true! [...] And now you're being a **patronizing jerk** too. [...]  I'm unsubscribing from this thread now. [...] I don't care enough to put up with your behavior [...] Please don't tag me in again.
 
Can't say I'll miss you.